clippy_lints/src/methods/unnecessary_map_or.rs

```diff
@@ -0,0 +1,132 @@
+            && let closure_body = cx.tcx.hir().body(map_closure.body)
+            && let closure_body_value = closure_body.value.peel_blocks()
+            && let ExprKind::Binary(op, l, r) = closure_body_value.kind
+            && let Some(param) = closure_body.params.first()
+            && let PatKind::Binding(_, hir_id, _, _) = param.pat.kind
+            // checking that map_or is one of the following:
+            // .map_or(false, |x| x == y)
+            // .map_or(false, |x| y == x) - swapped comparison
+            // .map_or(true, |x| x != y)
+            // .map_or(true, |x| y != x) - swapped comparison
+            && ((BinOpKind::Eq == op.node && !def_bool) || (BinOpKind::Ne == op.node && def_bool))
+            && let non_binding_location = if path_to_local_id(l, hir_id) { r } else { l }
+            && switch_to_eager_eval(cx, non_binding_location)
+            // xor, because if its both then thats a strange edge case and
+            // we can just ignore it, since by default clippy will error on this
+            && (path_to_local_id(l, hir_id) ^ path_to_local_id(r, hir_id))
+            && !is_local_used(cx, non_binding_location, hir_id)
+            && let typeck_results = cx.typeck_results()
+            && typeck_results.expr_ty(l) == typeck_results.expr_ty(r)
+    {
+        let wrap = variant.variant_name();
+        let comparator = op.node.as_str();
+
+        // we may need to add parens around the suggestion
+        // in case the parent expression has additional method calls,
+        // since for example `Some(5).map_or(false, |x| x == 5).then(|| 1)`
+        // being converted to `Some(5) == Some(5).then(|| 1)` isnt
+        // the same thing
+        let should_add_parens = get_parent_expr(cx, expr)
+            .is_some_and(|expr| expr.precedence().order() > i8::try_from(AssocOp::Equal.precedence()).unwrap_or(0));
+        (
+            format!(
+                "{}{} {comparator} {wrap}({}){}",
+                if should_add_parens { "(" } else { "" },
+                snippet(cx, recv.span, ".."),
+                snippet(cx, non_binding_location.span.source_callsite(), ".."),
+                if should_add_parens { ")" } else { "" }
+            ),
+            "standard comparison",
+        )
+    } else if !def_bool
+        && msrv.meets(msrvs::OPTION_RESULT_IS_VARIANT_AND)
+        && let Some(recv_callsite) = snippet_opt(cx, recv.span.source_callsite())
+        && let Some(span_callsite) = snippet_opt(cx, map.span.source_callsite())
+    {
+        let suggested_name = variant.method_name();
+        (
+            format!("{recv_callsite}.{suggested_name}({span_callsite})",),
+            suggested_name,
+        )
+    } else {
+        return;
+    };
+
+    span_lint_and_sugg(
+        cx,
+        UNNECESSARY_MAP_OR,
+        expr.span,
+        "this `map_or` is redundant",
+        format!("use {method} instead"),
+        sugg,
+        Applicability::MachineApplicable,
```

Comment:
I just noticed that this might not be machine applicable. Don't we need to check for `PartialEq` somewhere if we'll change it to `opt == <something>`?

Reply:
Yes. This is not a problem for an `Option<T>` as we know that `T` already implements `PartialEq` if the original code was `x.map_or(false, |x| == value)`, and thus `Option<T>` implements `PartialEq` as well. However, for `Result<T, E>`, `E` must implement `PartialEq` as well, or the comparison cannot be done at the `Result<T, E>` level.

For example, the following code will lead to an incorrect suggestion:

```rust
fn main() {
    struct S;
    let r: Result<i32, S> = Ok(3);
    let _ = r.map_or(false, |x| x == 7);
}
```

because the suggested replacement `r == Ok(7)` won't work.